incubator/validator-json/src/main/java/io/aklivity/zilla/runtime/validator/json/JsonValidator.java

```diff
@@ -95,31 +89,40 @@ protected boolean validate(
         return status;
     }
 
+    private JsonSchema supplySchema(
+        int schemaId)
+    {
+        return schemas.computeIfAbsent(schemaId, this::resolveSchema);
+    }
+
+    private JsonProvider supplyProvider(
+        int schemaId)
+    {
+        return providers.computeIfAbsent(schemaId, id -> createProvider(supplySchema(id)));
```

Comment:
Suggest we simplify by making `createProvider` take `schemaId` and calling `supplySchema` from `createProvider` implementation.